protocol/strict-derivation.md

```diff
@@ -0,0 +1,269 @@
+- _Steady Batch Derivation_ derives invalid singluar batches immediately as deposit-only
+blocks.
+- _Strict Batch Ordering_ requires batches within and across channels to be strictly ordered.
+
+The design space and some proposed solutions will be discussed, together with pratical implications
+for batcher implementations that have to satisfy the stricter rules.
+
+# Partial Span Batch Validity
+
+## Problem Statement + Context
+
+Before Holocene, Span Batch validity is handled atomically, i.e., either all singluar batches that
+can be derived from a span batch are valid, or if any of them is invalid (during batch queue
+checks), the whole span batch is dropped. This has negative implications for Fault Proofs: in a
+worst-case scenario, if the to-be-fault-proven block is the first batch in a span batch, to
+determine its validity at the batch queue stage, the validity of the full span batch has to be
+determined, as any later invalid singular batch would also invalidate all prior batches in the same
+span batch. So only invalidating the remaining span batch upon hitting an invalid derived singular
+batch vastly improves this scenario, since batch queue validity is then final for any singular
+batch, not just regular singular batches.
+
+In late 2023 while Delta was still in development, the problem of partial span batch validity was
+[already discussed](https://www.notion.so/oplabs/Handling-Invalid-Batches-48bb972368ce409c84bd9189eddd0577)
+and for simplicity it was decided to use the atomic validity model. However, above discussion
+focusses on implications of the atomic validity model for EL-sync. We're now concerned with behavior
+around CL client restarts and reorgs.
+
+### Mid-way restart
+
+One concern that was raised with partial span batch validity is how to deal with mid-way restarts,
+i.e. when the CL client restarts, loosing its internal derivation pipeline state while the safe head
+of the EL client is in the middle of a span batch. In particular, the correct behavior needs to
+cover the case where a L1 reorg happened in the meantime that reorged out the L1 block that
+contained the span batch.
+
+I fail to see currently how partial span batch validity makes this case worse or different. With
+atomic span batch validity, there's the pending-safe state of derived blocks that might be reorged
+out later by a backwards-invalidating singular batch derived from a span batch. So if we only
+forwards-invalidate, this becomes easier.
+
+And in any case, a restart or reorg needs to reconstruct the derivation pipeline state from the new
+(sync-start) L1 block, so I don't understand how reconstructing span batch derivation state is
+different from reconstructing the channel bank and batch queue in general. If a restart or reorg
+happens in the middle of deriving singular batches from a channel, we're in a similar situation that
+is solved by going back far enough for the sync-start.
+
+## Proposed Design
```

Comment:
I'm a fan of this design - adding a new optional stage here to incrementally stream single batches into the batch queue seems to make sense.

Beyond the outer checks on the span batch prefix, is the validity check on the single batches within the span batch a superset of existing single batch validity rules, or is it the same verbatim?

Reply:
Beyond that, would the new flow generally look like:
1. Process span batch prefix validity checks
2. `checkSingleBatch` on each single batch streamed over to the batch queue
?

If so, that's awesome for proofs on top of fixing the invalidation issue - getting rid of the overlapped blocks check within the batch validity checking logic should be huge on cycle count / IO. That check was taking ~30% of cycles (per call, done twice re: https://github.com/ethereum-optimism/optimism/issues/11317) last time we profiled kona.

Reply:
The singular batch checks are the same. The outer span batch checks are really just more of an early discard of the whole span batch because if these checks fail, there's no way that the singular batches pulled out of them could be valid.
The flow is correct. We could add _3. if the singular batch check fails, the remaining span batch is discarded_
This has to be because span batches don't encode the parent block hashes, so they are implicit. If any singular batch is invalid, it will be derived as a deposit-only block which will have a different block hash, so the implicit block hash must be assumed wrong.